presto-main/src/main/java/com/facebook/presto/sql/analyzer/FunctionArgumentCheckerForAccessControlUtils.java

```diff
@@ -0,0 +1,170 @@
+
+    private FunctionArgumentCheckerForAccessControlUtils() {}
+
+    // Returns whether function argument at `argumentIndex` for function `node` needs to be checked
+    // for column level access control.
+    // For e.g., consider SQL `transform(arr, col -> col.x)`
+    // Here, we only need to check for access of subfield `x` in column `arr` which is of type `Array<struct>`.
+    // So we can just parse lambda and ignore the first argument for access checks.
+    public static boolean isUnusedArgumentForAccessControl(FunctionCall node, int argumentIndex, ExpressionAnalyzer.Context context)
+    {
+        if (node.getName().equals(TRANSFORM)) {
+            checkState(node.getArguments().size() == 2);
+            return argumentIndex == 0;
+        }
+        return false;
+    }
+
+    // Parses arguments of function `node` which are a lambda expression, and returns a map
+    // of their lambda arguments to resolved subfield.
+    // For e.g., consider SQL `SELECT transform(arr, col -> col.x) FROM table`
+    // Return value = Map('col' -> ResolvedSubfield(table.arr))
+    public static Map<Identifier, ResolvedSubfield> getResolvedLambdaArguments(
+            FunctionCall node,
+            StackableAstVisitorContext<ExpressionAnalyzer.Context> context,
+            Map<NodeRef<Expression>, Type> expressionTypes)
+    {
+        ImmutableMap.Builder<Identifier, ResolvedSubfield> resolvedLambdaArguments = ImmutableMap.builder();
+        if (node.getName().equals(TRANSFORM)) {
+            checkState(node.getArguments().size() == 2);
+            if (!(node.getArguments().get(1) instanceof LambdaExpression)) {
+                return ImmutableMap.of();
+            }
+            Expression arrayExpression = node.getArguments().get(0);
+            LambdaExpression lambdaExpression = ((LambdaExpression) node.getArguments().get(1));
+            Optional<ResolvedSubfield> resolvedSubfield = resolveSubfield(arrayExpression, context, expressionTypes);
+            if (resolvedSubfield.isPresent()) {
+                resolvedLambdaArguments.put(
+                        lambdaExpression.getArguments().get(0).getName(),
+                        resolvedSubfield.get());
+            }
+        }
+        return resolvedLambdaArguments.build();
+    }
+
+    public static Optional<ResolvedSubfield> resolveSubfield(
+            Expression node,
+            StackableAstVisitorContext<ExpressionAnalyzer.Context> context,
+            Map<NodeRef<Expression>, Type> expressionTypes)
+    {
+        // If expression is nested with multiple dereferences and subscripts, we only look at the topmost one.
+        if (!isTopMostReference(node, context)) {
+            return Optional.empty();
+        }
+
+        Scope scope = context.getContext().getScope();
+        Expression childNode = node;
+        List<Subfield.PathElement> columnDereferences = new ArrayList<>();
+        while (true) {
```

Comment:
i think all the different cases here are a bit confusing.  Needs some comments, but I think it would be clearer if written as a visitor instead of a while loop with each case determining whether to continue the loop.  It would also be easy to introduce an infinite loop in this logic(continue gets called without the child nod getting updated to something that would break the loop eventually)

Reply:
I don't want to modify this logic in this PR. This is an existing logic moved from ExpressionAnalyzer to here, which is already tested. I can put up a todo to refactor it.

Reply:
Added comments

Reply:
got it.  I guess it's okay since the code already existed.  would be good to refactor though as a follow up.